From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Tue, 2 Jan 2024 10:00:00 +0100
Subject: [PATCH 1/5] Revert "(#1019): Fix small-caps and bump the version of Pango
 required to 1.50.0"

This reverts commit 73c1ee7519555aa90c1c45828ffaa2f573812864.

RHEL 9 doesn't provide Pango 1.50.0.

diff --git a/devel-docs/_build_dependencies.rst b/devel-docs/_build_dependencies.rst
index 1111111..2222222 100644
--- a/devel-docs/_build_dependencies.rst
+++ b/devel-docs/_build_dependencies.rst
@@ -19,7 +19,7 @@ minimum version is listed here; you may use a newer version instead.
 * `Freetype2 <https://gitlab.freedesktop.org/freetype/freetype>`_ 2.8.0
 * `GLib <https://gitlab.gnome.org/GNOME/glib/>`_ 2.50.0
 * `Libxml2 <https://gitlab.gnome.org/GNOME/libxml2>`_ 2.9.0
-* `Pango <https://gitlab.gnome.org/GNOME/pango/>`_ 1.50.0
+* `Pango <https://gitlab.gnome.org/GNOME/pango/>`_ 1.46.0
 
 **Optional dependencies:**
 
diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -56,7 +56,7 @@ freetype2_required     = '>= 20.0.14' # Corresponds to ft2_cmake_required
 glib_required          = '>= 2.50.0'
 harfbuzz_required      = '>= 2.0.0'
 libxml_required        = '>= 2.9.0'
-pango_required         = '>= 1.50.0'
+pango_required         = '>= 1.46.0'
 
 # Optional dependencies
 gdk_pixbuf_required    = '>= 2.20'
diff --git a/rsvg/Cargo.toml b/rsvg/Cargo.toml
index 1111111..2222222 100644
--- a/rsvg/Cargo.toml
+++ b/rsvg/Cargo.toml
@@ -27,15 +27,15 @@ gio = { name = "gio-2.0", version = "2.24" }
 glib = { name = "glib-2.0", version = "2.50" }
 harfbuzz = "2.0"
 libxml2 = { name = "libxml-2.0", fallback-names = ["libxml2"], version = "2.9" }
-pangocairo = "1.50"
+pangocairo = "1.46"
 
 [package.metadata.system-deps.'cfg(any(target_os = "macos", target_os = "windows"))']
 fontconfig = { version = "1.7", optional = true }
-pangoft2 = { version = "1.50", optional = true }
+pangoft2 = { version = "1.46", optional = true }
 
 [package.metadata.system-deps.'cfg(all(not(target_os = "macos"), not(target_os = "windows")))']
 fontconfig = { version = "1.7" }
-pangoft2 = { version = "1.50" }
+pangoft2 = { version = "1.46" }
 
 [features]
 avif = ["image/avif-native"]
diff --git a/rsvg/src/text.rs b/rsvg/src/text.rs
index 1111111..2222222 100644
--- a/rsvg/src/text.rs
+++ b/rsvg/src/text.rs
@@ -1399,7 +1399,9 @@ fn add_pango_attributes(
     font_desc.set_family(props.font_family.as_str());
     font_desc.set_style(pango::Style::from(props.font_style));
 
-    font_desc.set_variant(pango::Variant::from(props.font_variant));
+    // PANGO_VARIANT_SMALL_CAPS does nothing: https://gitlab.gnome.org/GNOME/pango/-/issues/566
+    // see below for using the "smcp" OpenType feature for fonts that support it.
+    // font_desc.set_variant(pango::Variant::from(props.font_variant));
 
     font_desc.set_weight(pango::Weight::from(props.font_weight));
     font_desc.set_stretch(pango::Stretch::from(props.font_stretch));
@@ -1422,6 +1424,13 @@ fn add_pango_attributes(
         attributes.push(pango::AttrInt::new_strikethrough(true).upcast());
     }
 
+    // FIXME: Using the "smcp" OpenType feature only works for fonts that support it.  We
+    // should query if the font supports small caps, and synthesize them if it doesn't.
+    if props.font_variant == FontVariant::SmallCaps {
+        // smcp - small capitals - https://docs.microsoft.com/en-ca/typography/opentype/spec/features_pt#smcp
+        attributes.push(pango::AttrFontFeatures::new("'smcp' 1").upcast());
+    }
+
     // Set the range in each attribute
 
     for attr in &mut attributes {

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 15 Aug 2024 14:50:22 +0200
Subject: [PATCH 2/5] Revert "Bump Cairo requirement to 1.18.0 so we can update
 gtk-rs"

This reverts commit ec5d747479f4098669bb93901f3008d412e3f62f.

RHEL 9 doesn't provide Cairo 1.18.0.

diff --git a/devel-docs/_build_dependencies.rst b/devel-docs/_build_dependencies.rst
index 1111111..2222222 100644
--- a/devel-docs/_build_dependencies.rst
+++ b/devel-docs/_build_dependencies.rst
@@ -15,7 +15,7 @@ minimum version is listed here; you may use a newer version instead.
 
 **Mandatory dependencies:**
 
-* `Cairo <https://gitlab.freedesktop.org/cairo/cairo>`_ 1.18.0 with PNG support
+* `Cairo <https://gitlab.freedesktop.org/cairo/cairo>`_ 1.17.0 with PNG support
 * `Freetype2 <https://gitlab.freedesktop.org/freetype/freetype>`_ 2.8.0
 * `GLib <https://gitlab.gnome.org/GNOME/glib/>`_ 2.50.0
 * `Libxml2 <https://gitlab.gnome.org/GNOME/libxml2>`_ 2.9.0
diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -49,7 +49,7 @@ py = import('python')
 python = py.find_installation()
 
 # Required versions for dependencies - Please update _build_dependencies.rst with version numbers if these change
-cairo_required         = '>= 1.18.0'
+cairo_required         = '>= 1.17.0'
 dav1d_required         = '>= 1.3.0'
 ft2_cmake_required     = '>= 2.8.0'   # Actual FreeType version required
 freetype2_required     = '>= 20.0.14' # Corresponds to ft2_cmake_required
diff --git a/rsvg/Cargo.toml b/rsvg/Cargo.toml
index 1111111..2222222 100644
--- a/rsvg/Cargo.toml
+++ b/rsvg/Cargo.toml
@@ -16,12 +16,12 @@ rustc-args = ["--cfg", "docsrs"]
 rustdoc-args = ["--cfg", "docsrs"]
 
 [package.metadata.system-deps]
-cairo = "1.18"
-cairo-gobject = "1.18"
-cairo-png = "1.18"
-cairo-pdf = { version = "1.18", optional = true }
-cairo-ps = { version = "1.18", optional = true }
-cairo-svg = { version = "1.18", optional = true }
+cairo = "1.17"
+cairo-gobject = "1.17"
+cairo-png = "1.17"
+cairo-pdf = { version = "1.17", optional = true }
+cairo-ps = { version = "1.17", optional = true }
+cairo-svg = { version = "1.17", optional = true }
 freetype2 = "20.0.14"
 gio = { name = "gio-2.0", version = "2.24" }
 glib = { name = "glib-2.0", version = "2.50" }
diff --git a/rsvg_convert/Cargo.toml b/rsvg_convert/Cargo.toml
index 1111111..2222222 100644
--- a/rsvg_convert/Cargo.toml
+++ b/rsvg_convert/Cargo.toml
@@ -13,9 +13,9 @@ rust-version.workspace = true
 autobins = false
 
 [package.metadata.system-deps]
-cairo-pdf = { version = "1.18", optional = true }
-cairo-ps = { version = "1.18", optional = true }
-cairo-svg = { version = "1.18", optional = true }
+cairo-pdf = { version = "1.17", optional = true }
+cairo-ps = { version = "1.17", optional = true }
+cairo-svg = { version = "1.17", optional = true }
 
 [dependencies]
 cairo-rs = { workspace = true, features = [

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 15 Aug 2024 17:17:11 +0200
Subject: [PATCH 3/5] Revert "add pdf1.7 as additional format"

This reverts commit c88987b77e8cdde51bfd4defbbce47c3b41221eb.

RHEL 9 doesn't provide Cairo 1.18.0.

diff --git a/rsvg_convert/src/main.rs b/rsvg_convert/src/main.rs
index 1111111..2222222 100644
--- a/rsvg_convert/src/main.rs
+++ b/rsvg_convert/src/main.rs
@@ -257,7 +257,6 @@ impl Surface {
         match format {
             Format::Png => Self::new_for_png(size, stream),
             Format::Pdf => Self::new_for_pdf(size, stream, None),
-            Format::Pdf1_7 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_7)),
             Format::Pdf1_6 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_6)),
             Format::Pdf1_5 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_5)),
             Format::Pdf1_4 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_4)),
@@ -490,7 +489,6 @@ impl std::fmt::Display for Output {
 enum Format {
     Png,
     Pdf,
-    Pdf1_7,
     Pdf1_6,
     Pdf1_5,
     Pdf1_4,
@@ -611,7 +609,6 @@ impl Converter {
                 }
 
                 Format::Pdf
-                | Format::Pdf1_7
                 | Format::Pdf1_6
                 | Format::Pdf1_5
                 | Format::Pdf1_4
@@ -808,8 +805,6 @@ fn build_cli() -> clap::Command {
         #[cfg(system_deps_have_cairo_pdf)]
         "pdf",
         #[cfg(system_deps_have_cairo_pdf)]
-        "pdf1.7",
-        #[cfg(system_deps_have_cairo_pdf)]
         "pdf1.6",
         #[cfg(system_deps_have_cairo_pdf)]
         "pdf1.5",
@@ -1086,7 +1081,6 @@ fn parse_args() -> Result<Converter, Error> {
         format_str,
         "png" => Format::Png,
         "pdf" => Format::Pdf,
-        "pdf1.7" => Format::Pdf1_7,
         "pdf1.6" => Format::Pdf1_6,
         "pdf1.5" => Format::Pdf1_5,
         "pdf1.4" => Format::Pdf1_4,
diff --git a/rsvg_convert/tests/rsvg_convert.rs b/rsvg_convert/tests/rsvg_convert.rs
index 1111111..2222222 100644
--- a/rsvg_convert/tests/rsvg_convert.rs
+++ b/rsvg_convert/tests/rsvg_convert.rs
@@ -151,16 +151,6 @@ fn output_format_pdf() {
         .stdout(file::is_pdf());
 }
 
-#[cfg(system_deps_have_cairo_pdf)]
-#[test]
-fn output_format_pdf_1_7() {
-    RsvgConvert::new_with_input("tests/fixtures/bug521-with-viewbox.svg")
-        .arg("--format=pdf1.7")
-        .assert()
-        .success()
-        .stdout(file::is_pdf().with_version("1.7"));
-}
-
 #[cfg(system_deps_have_cairo_pdf)]
 #[test]
 fn output_format_pdf_1_6() {

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 15 Aug 2024 17:18:10 +0200
Subject: [PATCH 4/5] Revert "add PDF version 1.6"

This reverts commit 166f74f593201fea7e95a8cd961cab88cf644bf7.

RHEL 9 doesn't provide Cairo 1.18.0.

diff --git a/rsvg_convert/Cargo.toml b/rsvg_convert/Cargo.toml
index 1111111..2222222 100644
--- a/rsvg_convert/Cargo.toml
+++ b/rsvg_convert/Cargo.toml
@@ -19,7 +19,6 @@ cairo-svg = { version = "1.17", optional = true }
 
 [dependencies]
 cairo-rs = { workspace = true, features = [
-    "v1_18",
     "v1_16",
     "pdf",
     "ps",
diff --git a/rsvg_convert/src/main.rs b/rsvg_convert/src/main.rs
index 1111111..2222222 100644
--- a/rsvg_convert/src/main.rs
+++ b/rsvg_convert/src/main.rs
@@ -257,7 +257,6 @@ impl Surface {
         match format {
             Format::Png => Self::new_for_png(size, stream),
             Format::Pdf => Self::new_for_pdf(size, stream, None),
-            Format::Pdf1_6 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_6)),
             Format::Pdf1_5 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_5)),
             Format::Pdf1_4 => Self::new_for_pdf(size, stream, Some(cairo::PdfVersion::_1_4)),
             Format::Ps => Self::new_for_ps(size, stream, false),
@@ -489,7 +488,6 @@ impl std::fmt::Display for Output {
 enum Format {
     Png,
     Pdf,
-    Pdf1_6,
     Pdf1_5,
     Pdf1_4,
     Ps,
@@ -608,12 +606,7 @@ impl Converter {
                     )
                 }
 
-                Format::Pdf
-                | Format::Pdf1_6
-                | Format::Pdf1_5
-                | Format::Pdf1_4
-                | Format::Ps
-                | Format::Eps => {
+                Format::Pdf | Format::Pdf1_5 | Format::Pdf1_4 | Format::Ps | Format::Eps => {
                     // These surfaces require units in points
                     unit = LengthUnit::Pt;
 
@@ -805,8 +798,6 @@ fn build_cli() -> clap::Command {
         #[cfg(system_deps_have_cairo_pdf)]
         "pdf",
         #[cfg(system_deps_have_cairo_pdf)]
-        "pdf1.6",
-        #[cfg(system_deps_have_cairo_pdf)]
         "pdf1.5",
         #[cfg(system_deps_have_cairo_pdf)]
         "pdf1.4",
@@ -1081,7 +1072,6 @@ fn parse_args() -> Result<Converter, Error> {
         format_str,
         "png" => Format::Png,
         "pdf" => Format::Pdf,
-        "pdf1.6" => Format::Pdf1_6,
         "pdf1.5" => Format::Pdf1_5,
         "pdf1.4" => Format::Pdf1_4,
         "ps" => Format::Ps,
diff --git a/rsvg_convert/tests/rsvg_convert.rs b/rsvg_convert/tests/rsvg_convert.rs
index 1111111..2222222 100644
--- a/rsvg_convert/tests/rsvg_convert.rs
+++ b/rsvg_convert/tests/rsvg_convert.rs
@@ -151,16 +151,6 @@ fn output_format_pdf() {
         .stdout(file::is_pdf());
 }
 
-#[cfg(system_deps_have_cairo_pdf)]
-#[test]
-fn output_format_pdf_1_6() {
-    RsvgConvert::new_with_input("tests/fixtures/bug521-with-viewbox.svg")
-        .arg("--format=pdf1.6")
-        .assert()
-        .success()
-        .stdout(file::is_pdf().with_version("1.6"));
-}
-
 #[cfg(system_deps_have_cairo_pdf)]
 #[test]
 fn output_format_pdf_1_5() {

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 15 Aug 2024 15:34:44 +0200
Subject: [PATCH 5/5] Downgrade minimum required Meson version to 0.63.3

To ensure compat with RHEL 9.

diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -1,7 +1,7 @@
 project('librsvg',
         'c',
         version: '2.59.90', # Keep this in sync with Cargo.toml, doc/librsvg.toml, rsvg/Cargo.toml
-        meson_version: '>= 1.3.0',
+        meson_version: '>= 0.63.3',
         default_options: [
         ]
 )
@@ -347,7 +347,7 @@ if host_system in ['windows', 'linux']
   endif
 
   foreach i: native_libs.stdout().split()
-    if 'msvcrt' in i
+    if i.contains('msvcrt')
       if i.endswith('.lib') # pre Rust 1.83
         default_overrides = [
           'b_vscrt=md@0@'.format(i.substring(0, -4).split('msvcrt')[1])
@@ -447,7 +447,7 @@ else
 endif
 
 cargo_dylib_prefix = lib_prefix
-if cc.has_define('__MINGW32__')
+if cc.get_define('__MINGW32__') != ''
   cargo_dylib_prefix = ''
 endif
 
